### ISSUE TITLE
Slow volume attach for single ReadOnlyMany volume across multiple nodes

### ISSUE THREAD

**Issue by verult** [kubernetes/kubernetes#73972]
**What happened**: A GCE PD volume is used in read-only mode (through a PVC & PV pair in ReadOnlyMany mode) by over 100 pods running across tens of nodes. Volume attachment sometimes takes over 30min for some of these pods.

**What you expected to happen**: Attaching should take seconds instead of minutes.

**How to reproduce it (as minimally and precisely as possible)**:
* Provision a read-only GCE PD, and the associated PVC & PV.
* Run a Job with parallelism of 100+ pods in a cluster with tens of nodes (we tested with 50+), and use the PD inside all pods.

**Anything else we need to know?**: Generally attach/detach calls are parallelized (through OperationExecutor), but calls with the same volume name are serialized. This was done as a way to prevent races between attach and detach calls for the same volume. However, it's sufficient to provide this guarantee for each node, as long as the volume is Read[Write|Only]Many.

https://github.com/kubernetes/kubernetes/blob/f96849981203ea506ea8fa3f88c173d36cf688aa/pkg/volume/util/operationexecutor/operation_executor.go#L607-L608

**Environment**:
- Kubernetes version (use `kubectl version`): Tested with multiple 1.10 patch releases (including 1.10.6 and 1.10.9)
- Cloud provider or hardware configuration: GCP

/sig storage
/priority critical-urgent
/assign
/cc @davidz627 @jingxu97 @msau42 @saad-ali

**Comment by davidz627** [comment]
Even though the calls are serialized I wouldn't expect `100` serialized attaches to take `>30mins`. Could there be something else at play here? Are we hitting some sort of error case and exponential backoff-ing when we shouldn't be? I would assume that follow-on attaches are basically a no-op and completed very quickly the first time

**Comment by davidz627** [comment]
Looking here the re-attach is basically a no-op: https://github.com/kubernetes/kubernetes/blob/2981fb7a0171798184b4f860b07870de0ff7b234/pkg/volume/gcepd/attacher.go#L96
although this call in order to determine whether the disk is already attached is pretty hefty and i've seen issues with it in the past: https://github.com/kubernetes/kubernetes/blob/2981fb7a0171798184b4f860b07870de0ff7b234/pkg/volume/gcepd/attacher.go#L86
since it gets instances from the cloud I've observed quota issues when this is hit repeatedly in a short period of time in the past.

If we parallelize the attaches the quota issue there becomes an even bigger problem with concurrent attaches (not necessarily related to this issue but thought I'd bring it up).

**Comment by gnufied** [comment]
Yeah, it would be interesting to see controller-manager logs for just that particular volume when this happened. It should also be possible to grab controller-manager operation metrics which should tell us if attach itself is taking time or simply attach not being called. This is the time when I guess - total operation metric would have been more useful.

**Comment by verult** [comment]
@davidz627 in the case we are seeing, because there's a large number of nodes, the number of replicas scheduled per node is likely pretty low, so re-attach is not the common case. I agree though that an average of 18s per attach call still doesn't feel quite right and there might be another error at play here. Per @gnufied 's suggestions I'll take a look at operation latency.

Regardless of whether it's the sole cause, the fact that attach calls are serialized even if they are for different nodes seems like a scalability bug to me, especially for Read[Write|Only]Many volumes

**Comment by droyo** [comment]
Looking at the `DiskIsAttached` method, it looks up the instance and searches its `disk` member.

In the GCE API, the [`disks.get`](https://cloud.google.com/compute/docs/reference/rest/v1/disks/get) method returns an object with a `users` array containing all instances that have the disk attached. A sensible fix would be to call this method instead and cache the results.

**Comment by verult** [comment]
@droyo sorry I don't quite follow - why is `DiskIsAttached` involved? Is your suggest an attempt to reduce per-attach latency?

@saad-ali and @msau42 proposed a fix of adding an extra `node` parameter to identify each attach goroutine, which is only set if the volume is Read*Many.

**Comment by jingxu97** [comment]
I think @droyo suggests to change the implementation of https://github.com/kubernetes/kubernetes/blob/master/pkg/cloudprovider/providers/gce/gce_disks.go#L580
Currently, it tries to getinstances and check the disk from the instance. The other option is to use getdisk API to verify whether disk is attached to the node. I am not sure how much performance differences between these two.

**Comment by msau42** [comment]
The problem we see in this issue is related to attach though, not verify attached.  Are you thinking that too many verify attach calls is slowing down attach too?

**Comment by jingxu97** [comment]
Attach call first call DiskIsAttach to see whether it is already attached or not.

**Comment by gnufied** [comment]
If we do plan to serialize operations on node+volume combination, we also have to fix https://github.com/kubernetes/kubernetes/blob/master/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L264 which will again try to serialize operations based on volume name.

**Comment by verult** [comment]
An update on the fix to parallelize by adding a node parameter - it's simple to figure out the access mode for PVCs, but harder for inline volumes. I'm not aware of a way for inline volumes to specify access modes, and "readonly" (which many volume sources have) is not equivalent to "available to multiple modes".

**Comment by verult** [comment]
Option 1: have a util function called `isAcessModeMany(spec)` or something like that, where for each volume plugin, if we don't currently have the knowledge of the plugin's behavior w.r.t the availability to multiple nodes, we default to `false` for the time being, and ask each plugin owner to set it appropriately if they want to take advantage of this parallelized attach.

Option 2: Don't have the Read*Many check; parallelize for all attaches. Sounds risky, but I don't know the details of our attach logic well enough to say whether we'd end up with races.

**Comment by msau42** [comment]
Option 2 sounds very risky.

Does read-only many even work for inline volumes?

**Comment by droyo** [comment]
Sorry for not being clear in my earlier comment. My comment was in response to @davidz627 's observation that `DiskIsAttached` is the heaviest part of re-attaching a disk that is already attached.

My point was that if you have one disk and many GCE instances, it is much faster to use `disks.get` to get the disk and check all instances against its `users` array, than making a separate API call for each instance. It will also keep you under the GCE read API quota more effectively. It is a good optimization when the number of pods using the volume is larger than the number of nodes in the cluster.

It does not improve the flow where the disk needs to be attached to a node. The parallelization efforts being discussed now will help with that.

**Comment by verult** [comment]
@droyo thanks for the clarification! I can make a separate PR to improve that.

@msau42 I'll play around with inline volumes today to see if it works, but your question also leads to:
Option 3: Only make the fix work for PVCs, and broadcast this behavior to sig storage

**Comment by verult** [comment]
@msau42 Read-only many does work for inline GCE PD volumes

**Comment by verult** [comment]
@msau42 @saad-ali and I discussed for a bit offline.

Our proposal is to add the `node` parameter for *all* volume types, not just Read*Many. The potential races for volumes not allowing multi-node attach should be handled by `IsMultiAttachForbidden()` in reconciler.
https://github.com/kubernetes/kubernetes/blob/07da603b59052036c00b4cb369b837c7336b50ee/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L271

and if it doesn't handle it properly today, we should modify that instead of having custom logic in OperationExecutor.

**Comment by gnufied** [comment]
That could cause an detach and a attach to happen at the same time, which may or may not be safe depending on volume type.  Could this cause volume types that don't have any inbuilt fencing to be available on 2 nodes concurrently? We have seen file system corruption in case it does happen.

**Comment by gnufied** [comment]
On second thought - I think we are assuming if detach has not finished yet then actual attach call won't be made until volume is removed from actual state of the world(if volume does not support multi-attach). So we should be safe? 

I see two problems with that:

1. We should review and make sure that each volume type emits correct access mode. 
2. Second problem I see is - mutliAttach check is entirely based on cached information within attach/detach controller. So is it possible that - two pods try to use a volume X but neither has succeeded yet but they are concurrently trying to mount same volume? multiAttach check won't prevent that - I think.

**Comment by verult** [comment]
Linking past discussions around multi-attach here: https://github.com/kubernetes/kubernetes/issues/40669 and https://github.com/kubernetes/kubernetes/pull/40148#pullrequestreview-18685838 .

Context for those who didn't really know how `IsMultiAttach()` worked (like me :) ):
* For in-line volumes, it encodes a check on a per-volume-plugin basis. However this doesn't cover all volume plugins today so we run a risk of breaking those volume plugins if inline volumes are used.
* For PV/PVCs, it relies on the access mode, and allows multi-attach if it's Read*Many. This is why @gnufied mentioned PV/PVC's access modes must be correct.

* Regarding (1): I believe the user should ultimately be responsible for setting the correct access mode in the PV/PVC. To minimize the risk of accidental data corruption caused by setting the wrong access mode (and erroneously triggering multi-attach), volume plugins should be responsible for verifying this access mode against the underlying storage.
* Regarding (2): Just to verify my understanding - the call to `IsMultiAttachForbidden()` relies on volume spec inside the DesiredStateOfWorld. When two pods try to use volume X, the reconcile loop only checks (a) whether it has already been attached (which it isn't), and (b) whether an attach operation is pending. To prevent attaching in parallel, the `IsOperationPending()` check could be changed to do something different depending on whether multi-attach is allowed. If it is, check if an operation with the same (volumeName, nodeName) is pending; otherwise, check for operations with the same volumeName only.
* If attaches were always serialized within the same volume, does that mean we've never triggered a multi-attach scenario with k8s in production, i.e. `IsMultiAttachForbidden()` was never used against the race condition it's supposed to guard against? If this is the case, we definitely need (a) stronger integration & e2e tests and (b) broadly communicate this change to SIG-storage, especially to volume plugin owners, and have them verify that `IsMultiAttachForbidden()` is correct for their plugin (that inline volumes are handled correctly, and that access modes are verified).
* CSI drivers must also verify access modes. Do we need to do extra work for this in our k8s-CSI integration?

**Comment by verult** [comment]
Jumping on this bug again. @gnufied curious about your thoughts on my last comment

**Comment by davidz627** [comment]
/remove-lifecycle stale
has this been resolved?

**Comment by verult** [comment]
No it has not

**Comment by verult** [comment]
Answering my own questions [above](https://github.com/kubernetes/kubernetes/issues/73972#issuecomment-466155706) :

For (3), `IsMultiAttachForbidden()` probably wasn't designed with parallelism in mind, i.e. its original intention was probably not to guard against races.

I confirmed by inspection that the race @gnufied mentioned in his #2 point does exist:
https://github.com/kubernetes/kubernetes/blob/7bac6ca73ad6fa1b0b637e801a524d6a1df663ea/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L272
`GetNodesForAttachedVolume()` returns a list of attached nodes for the volume in ASW, and this information is updated at the end of `Attach()`, which will run in parallel.
 For (2), putting multi-attach logic in `IsOperationPending()` or anywhere else in the operation executor could get messy. I'd like to avoid it if possible, and instead fix the race with the logic that already exists in reconciler.

IMO the multi-attach logic should check the list of volumes *intending* to attach (i.e. read from DSW), rather than those that are *actually* attached.

**Comment by verult** [comment]
I take back what I said here:
> IMO the multi-attach logic should check the list of volumes *intending* to attach (i.e. read from DSW), rather than those that are *actually* attached.

Checking the intention isn't enough. In a particular iteration of the reconciler loop, if the DSW has multiple attaches for the same volume on different nodes, there isn't enough information to decide whether the current operation in question during this iteration should proceed. There are really 3 states w.r.t attaching:
1. attach intended
1. attach in progress
1. attach complete (actually attached)

At every iteration, the reconciler needs to know that if it were to issue an attach, no other attaches for the volume in question ever completes before this attach it's about to issue. If it's not sure about this, it shouldn't issue the attach.

It will be sure if:
1. there is no attach operation pending (attach in progress). Because attach is only issued by the reconciler, which executes serially, there will never be a parallel attach for the same volume until the current attach it's about to issue completes.
1. the volume isn't actually attached anywhere (attach complete).

Note the above checks must be executed in this order. Reversing the two introduces a race condition.

Given this, I take back another statement:
> I confirmed by inspection that the race @gnufied mentioned in his #2 point does exist

It actually works fine today because the code does exactly what I described:
https://github.com/kubernetes/kubernetes/blob/7bac6ca73ad6fa1b0b637e801a524d6a1df663ea/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L264
then
https://github.com/kubernetes/kubernetes/blob/7bac6ca73ad6fa1b0b637e801a524d6a1df663ea/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L272

And replying to @gnufied 's comment here:
> On second thought - I think we are assuming if detach has not finished yet then actual attach call won't be made until volume is removed from actual state of the world(if volume does not support multi-attach). So we should be safe?
This is my understanding as well. `IsOperationPending()` doesn't key on the operation name.

**Comment by verult** [comment]
This might not be right:

https://github.com/kubernetes/kubernetes/blob/81e9f21f832f88422f1ccf5b8aa90de7cf822132/pkg/controller/volume/attachdetach/reconciler/reconciler.go#L162

To be safe, we should return multi-attach forbidden if there exists ReadWriteOnce in the access mode list

**Comment by tedyu** [comment]
I think @verult is right.
Created #85166

**Comment by liggitt** [comment]
lowering priority, this doesn't appear to be a regression or data loss issue
